### PR TITLE
runtime-rs: add oci hook support

### PR DESCRIPTION
**Pull request by YushuoEdge**
In this pr, we are going to support OCI Hooks in runtime-rs. Besides `prestart` / `poststart` / `poststop` / `create_runtime hooks`, which are supported in runtime-go, we also support `create_container` / `start_container` hooks in this pr.

Description of  OCI hooks in OCI spec: https://github.com/opencontainers/runtime-spec/blob/main/runtime.md#lifecycle

Fixes: #5787

Signed-off-by: Yushuo <y-shuo@linux.alibaba.com>

### CODE REVIEW

File: src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+        .stderr(Stdio::piped())
+        .spawn()?;
+
+    let state = serde_json::to_string(st)?;
+    let join_handle = tokio::spawn(async move {
+        // write state to stdin of child process
+        if let Some(mut stdin) = child.stdin.take() {
+            match stdin.write_all(state.as_bytes()).await {
+                Ok(_) => {}
+                Err(e) => {
```
Comment (Tim-0731-Hzt) [inline src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:101]: Should this return an error?

File: src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+    PostStart,
+    PostStop,
+}
+
+// execute hooks in oci::Hooks according to hook types
+pub async fn run_hooks(spec: &oci::Spec, st: &oci::State, hk_type: HookType) -> Result<()> {
+    if let Some(spec_hooks) = &spec.hooks {
+        let hooks = match hk_type {
+            HookType::PreStart => {
+                if !spec_hooks.prestart.is_empty() {
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:27]: I think this log and belows have no meaning, so delete them?

File: src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+                    .ok_or_else(|| anyhow!("hook exit status has no code"))?;
+                if exit_code != 0 {
+                    error!(
+                        sl!(),
+                        "hook {} exit code: {}, error message: {}",
+                        &binary_path.display(),
+                        exit_code,
+                        err
+                    );
+                    return Err(anyhow!(nix::Error::UnknownErrno));
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:146]: Is it better to return more detailed error like lines 152-156?

File: src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+
+// check and reserve valid environment variables
+// invalid env var may cause panic, refer to https://doc.rust-lang.org/std/env/fn.set_var.html#panics
+// key should not:
+//  * contain NUL character '\0'
+//  * contain ASCII equal sign '='
+//  * be empty
+// value should not:
+//  * contain NUL character '\0'
+fn valid_env(e: &str) -> Option<(&str, &str)> {
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:174]: Agent has the same code, and should use a common function laid in `src/libs`.

But you can leave it in another PR.

File: src/runtime-rs/crates/runtimes/src/manager.rs [hunk 5]
```diff
@@ -111,6 +117,12 @@ impl RuntimeHandlerManagerInner {
             .await
             .context("init runtime handler")?;
 
+        // CreateRuntime Hooks:
+        // * should be run in runtime namespace
+        // * should be run when creating the runtime
+        // * spec details: https://github.com/opencontainers/runtime-spec/blob/main/config.md#createruntime-hooks
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/src/manager.rs:123]: Don't use `main`/`master` in GitHub URL, use tag/commit instead.

File: src/runtime-rs/crates/runtimes/src/manager.rs [hunk 6]
```diff
@@ -205,7 +218,22 @@ impl RuntimeHandlerManager {
             let bundler_path = format!("{}/{}", req.bundle, oci::OCI_SPEC_CONFIG_FILE_NAME);
             let spec = oci::Spec::load(&bundler_path).context("load spec")?;
 
-            self.try_init_runtime_instance(&spec, &req.options)
+            // Prestart Hooks [DEPRECATED in newest oci spec]:
+            // * should be run in runtime namespace
+            // * should be run after "start", but if Prestart Hook and CreateRuntime Hook are both supported
+            //   it should be run before CreateRuntime Hook
+            // * spec details: https://github.com/opencontainers/runtime-spec/blob/main/config.md#prestart
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/src/manager.rs:225]: Ditto

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs [hunk 7]
```diff
@@ -89,6 +91,22 @@ impl ContainerManager for VirtContainerManager {
                 let c = containers
                     .remove(container_id)
                     .ok_or_else(|| Error::ContainerNotFound(container_id.to_string()))?;
+
+                // Poststop Hooks:
+                // * should be run in runtime namespace
+                // * should be run after the container is deleted but before delete operation returns
+                // * spec details: https://github.com/opencontainers/runtime-spec/blob/main/config.md#poststop
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:98]: Ditto

File: src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+        }
+
+        // read info from stdout of child process
+        if let Some(stdout) = child.stdout.as_mut() {
+            let mut out = String::new();
+            match stdout.read_to_string(&mut out).await {
+                Ok(_) => {
+                    info!(sl!(), "child stdout: {:?}", out.as_str());
+                }
+                Err(e) => {
```
Comment (Tim-0731-Hzt) [inline src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:114]: same as above

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs [hunk 9]
```diff
@@ -192,6 +210,22 @@ impl ContainerManager for VirtContainerManager {
             .get(container_id)
             .ok_or_else(|| Error::ContainerNotFound(container_id.clone()))?;
         c.start(process).await.context("start")?;
+
+        // Poststart Hooks:
+        // * should be run in runtime namespace
+        // * should be run after user-specific command is executed but before start operation returns
+        // * spec details: https://github.com/opencontainers/runtime-spec/blob/main/config.md#poststart
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:217]: Ditto

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs [hunk 10]
```diff
@@ -89,6 +91,22 @@ impl ContainerManager for VirtContainerManager {
...
                     .remove(container_id)
                     .ok_or_else(|| Error::ContainerNotFound(container_id.to_string()))?;
+
+                // Poststop Hooks:
+                // * should be run in runtime namespace
+                // * should be run after the container is deleted but before delete operation returns
+                // * spec details: https://github.com/opencontainers/runtime-spec/blob/main/config.md#poststop
+                let c_spec = c.spec().await;
+                let state = oci::State {
+                    version: String::from("ociVersion"),
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:101]: The `version` should a real version number, but not the `ociVersion` itself.

```json
    "state": {
        "ociVersion": "0.2.0",
        "id": "oci-container1",
        "status": "creating",
        "pid": 4422,
        "bundle": "/containers/redis",
        "annotations": {
            "myKey": "myValue"
        }
    }
```

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs [hunk 11]
```diff
@@ -192,6 +210,22 @@ impl ContainerManager for VirtContainerManager {
...
             .ok_or_else(|| Error::ContainerNotFound(container_id.clone()))?;
         c.start(process).await.context("start")?;
+
+        // Poststart Hooks:
+        // * should be run in runtime namespace
+        // * should be run after user-specific command is executed but before start operation returns
+        // * spec details: https://github.com/opencontainers/runtime-spec/blob/main/config.md#poststart
+        let c_spec = c.spec().await;
+        let state = oci::State {
+            version: String::from("ociVersion"),
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:220]: Ditto

File: src/runtime-rs/crates/runtimes/common/Cargo.toml [hunk 12]
```diff
@@ -26,3 +26,4 @@ agent = { path = "../../agent" }
 kata-sys-util = { path = "../../../../libs/kata-sys-util" }
 kata-types = { path = "../../../../libs/kata-types" }
 oci = { path = "../../../../libs/oci" }
+logging = { path = "../../../../libs/logging"}
```
Comment (Tim-0731-Hzt) [inline src/runtime-rs/crates/runtimes/common/Cargo.toml:29]: add a new line after this

File: src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs (hunk 2 above)
Comment (YushuoEdge) [reply src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:27]: We might need some log to indicate that the hook is running?

File: src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+            Err(e) => Err(anyhow!(
+                "wait child process error: {} {}",
+                e,
+                e.raw_os_error().unwrap()
+            )),
+        }
+    });
+
+    match tokio::time::timeout(execution_timeout, join_handle).await {
+        Ok(r) => r.unwrap(),
```
Comment (Tim-0731-Hzt) [inline src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:161]: how about make it to `Ok(r) => r.context("context")?`

File: src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs (hunk 3 above)
Comment (YushuoEdge) [reply src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:146]: The problem here is that there may be many kinds of errors, and it is difficult for us to determine which one.

File: src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+            HookType::PostStop => {
+                if !spec_hooks.poststop.is_empty() {
+                    info!(sl!(), "poststop hook");
+                }
+                &spec_hooks.poststop
+            }
+        };
+
+        for h in hooks.iter() {
+            run_hook(h, st).await?;
```
Comment (Tim-0731-Hzt) [inline src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:53]: suggest adding `.context()`

File: src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+
+    Ok(())
+}
+
+// execute a specific hook, regardless of hook type
+pub async fn run_hook(h: &oci::Hook, st: &oci::State) -> Result<()> {
+    info!(sl!(), "runtime action: execute hook");
+
+    // check whether the binary file called by the hook exists
+    let binary_path = PathBuf::from(h.path.as_str()).canonicalize()?;
```
Comment (Tim-0731-Hzt) [inline src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:65]: suggest adding `.context()`

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs [hunk 16]
```diff
@@ -89,6 +91,22 @@ impl ContainerManager for VirtContainerManager {
...
+                let c_spec = c.spec().await;
+                let state = oci::State {
+                    version: String::from("ociVersion"),
+                    id: c.container_id.to_string(),
+                    status: oci::ContainerState::Stopped,
+                    pid: c.pid().await as i32,
+                    bundle: c.config().await.bundle.clone(),
+                    annotations: c_spec.annotations.clone(),
+                };
+                run_hooks(&c_spec, &state, HookType::PostStop).await?;
```
Comment (Tim-0731-Hzt) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:108]: suggest adding `.context("run post stop")`

File: src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs (hunk 1 above)
Comment (YushuoEdge) [reply src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:101]: We probably don't need to return error here. I think it is best not to hinder the continued operation of the runtime because of the execution of the hook. But the log type here may indeed need to be changed.

File: src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs (hunk 3 above)
Comment (Ji-Xinyou) [reply src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:146]: The exitcode should be a Kernel Errno, you can convert that to Error. I think libc or nix can do that.

Comment (YushuoEdge) [reply src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:146]: I'm not sure whether it's a Kernel Errno at all here, and I need sometime to confirm ...

File: src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs (hunk 4 above)
Comment (YushuoEdge) [reply src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs:174]: I have found that there is also some code implementing `execute_hooks` in `src/libs`:
https://github.com/kata-containers/kata-containers/blob/main/src/libs/kata-sys-util/src/hooks.rs

Maybe I will use that code to refactor this PR. Or before that, also refactor the agent's hook execution with the same code (:

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs [hunk 17]
```diff
@@ -192,6 +210,22 @@ impl ContainerManager for VirtContainerManager {
...
+        // Poststart Hooks:
+        // * should be run in runtime namespace
+        // * should be run after user-specific command is executed but before start operation returns
+        // * spec details: https://github.com/opencontainers/runtime-spec/blob/main/config.md#poststart
+        let c_spec = c.spec().await;
+        let state = oci::State {
+            version: String::from("ociVersion"),
+            id: c.container_id.to_string(),
+            status: oci::ContainerState::Running,
+            pid: c.pid().await as i32,
```
Comment (bergwolf) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:223]: As per https://github.com/kata-containers/kata-containers/issues/5321, docker relies on the pid of the container to find out the container namespace. So please let's set state.pid to the vmm thread id as the vmm is the only process that is inside the container netns on the host.

In the meantime, it means that we need to run the prestart hooks after the vmm is launched.

**Comment by bergwolf** [comment]
I would like to ask for the feature to be discussed and design decisions to be made before merging the PR. See  https://github.com/kata-containers/kata-containers/issues/5787#issuecomment-1331748449

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs (hunk 17 above)
Comment (YushuoEdge) [reply src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:223]: Does it mean that state.pid of all OCI hooks should be set to vmm thread/process id?

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs [hunk 18]
```diff
@@ -192,6 +210,22 @@ impl ContainerManager for VirtContainerManager {
             .get(container_id)
             .ok_or_else(|| Error::ContainerNotFound(container_id.clone()))?;
         c.start(process).await.context("start")?;
+
+        // Poststart Hooks:
+        // * should be run in runtime namespace
```
Comment (Tim-0731-Hzt) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:215]: IIUC, runtime namespace means container namespace, do you need to enter the network namespace?  like `netns::NetnsGuard::new` Do I missing something?

Comment (YushuoEdge) [reply src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:215]: According to the spec, `runtime namespace` is not `container namespace`.

Comment (Tim-0731-Hzt) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:215]: IIUC, runtime namespace is container namespace, should enter network namespace?

Comment (Tim-0731-Hzt) [reply src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:215]: Thanks @YushuoEdge , I see that

**Comment by bergwolf** [comment]
@YushuoEdge Thanks for renewing the PR! Could you please add a design doc in the `docs` directory to explain the hook handling logic? The table in https://github.com/kata-containers/kata-containers/issues/5787#issuecomment-1331748449 can be included there as well.

**Comment by YushuoEdge** [comment]
@bergwolf The design document has been added, PTAL！

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs (hunk 17 above)
Comment (YushuoEdge) [reply src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs]: Updated!

File: src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs [hunk 19]
```diff
@@ -151,6 +190,16 @@ impl Sandbox for VirtSandbox {
...
 
+        // execute pre-start hook functions, including Prestart Hooks and CreateRuntime Hooks
+        let (prestart_hooks, create_runtime_hooks) = match spec.hooks.as_ref() {
+            Some(hooks) => (hooks.prestart.clone(), hooks.create_runtime.clone()),
+            None => (Vec::new(), Vec::new()),
+        };
+        self.execute_prestart_hooks(&prestart_hooks, &create_runtime_hooks, state)
+            .await?;
+
+        // TODO: if prestart_hooks is empty, rescan the network endpoints
```
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs:201]: Why leave this comment?

Comment (YushuoEdge) [reply src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs:201]: Sorry, it should be "is not". I will fix it.
This comment is related to https://github.com/kata-containers/kata-containers/pull/5972. 
As networks may be configured through `Prestart Hook`, so we should rescan the network if `prestart_hook` is not empty. Hotplug network endpoints has not supported yet and the rescan operation relies on that capability,  so I leave this comment as a TODO action.

File: docs/design/hooks-handling.md [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+guest_hook_path="/usr/share/hooks"
+```
++ In guest rootfs, prestart-hook is stored in `/usr/share/hooks/prestart/prestart-hook`.
+
+## Execution
+The table below summarized when and where those different hooks will be executed in Kata Containers:
+
+| Hook Name | Hook Type | Hook Path | Exec Place | Exec Time |
+|---|---|---|---|---|
+| `Prestart(deprecated)` | OCI hook | host runtime namespace | host runtime namespace | After VM is started, before container is created. |
```
Comment (Tim-0731-Hzt) [inline docs/design/hooks-handling.md:50]: I am confused about this, if it is deprecated, why do we still implement it?

Comment (liubin) [reply docs/design/hooks-handling.md:50]: Users may still use them.

Comment (bergwolf) [reply docs/design/hooks-handling.md:50]: It is deprecated by the spec, but there are still many users so we have to support it.

File: docs/design/hooks-handling.md [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+| `StartContainer` | OCI hook | guest container namespace | guest container namespace | After container is created, before container is started. |
+| `Poststart` | OCI hook | host runtime namespace | host runtime namespace | After container is started, before start operation returns. |
+| `Poststop` | OCI hook | host runtime namespace | host runtime namespace | After container is deleted, before delete operation returns. |
+| `Guest Prestart` | Kata hook | guest agent namespace | guest agent namespace | During start operation, before container command is executed. |
+| `Guest Poststart` | Kata hook | guest agent namespace | guest agent namespace | During start operation, after container command is executed, before start operation returns. |
+| `Guest Poststop` | Kata hook | guest agent namespace | guest agent namespace | During delete operation, after container is deleted, before delete operation returns. |
+
++ `Hook Path` specifies where hook's path be resolved.
++ `Exec Place` specifies in which namespace those hooks can be executed.
++ `Exec Time` specifies at which time point those hooks can be executed.
```
Comment (bergwolf) [inline docs/design/hooks-handling.md:62]: Please explain the design choice of  the `CreateContainer` hooks as in https://github.com/kata-containers/kata-containers/issues/5787#issuecomment-1331748449 too. We need to clearly document the inconsistent OCI-hook behavior that we choose to implement.

> One design choice that we need to make is that for CreateContainer hooks, OCI requires to run them inside the container namespace but the hook executable path is in the host runtime, which is a non-starter for VM-based containers. So I propose that we keep running it in the host vmm namespace. Other OCI hooks are found and run in the proper namespace as required by the OCI runtime.

File: src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs [hunk 22]
```diff
@@ -119,11 +120,49 @@ impl VirtSandbox {
 
         Ok(resource_configs)
     }
+
+    async fn execute_prestart_hooks(
```
[the remark this answers is not in the document]
Comment (YushuoEdge) [reply src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs:124]: I wrote this function just for binding this two kinds of hooks like `PrestartHookFunc` in runtime-go.  `prestart_hooks` and `create_runtime_hooks` refers to two specific kinds of hooks, while `execute_prestart_hooks` refers to execution time point.
But I agree that this name should be changed for understanding (:

File: docs/design/hooks-handling.md (hunk 21 above)
Comment (YushuoEdge) [reply docs/design/hooks-handling.md:63]: update!

File: src/libs/protocols/protos/oci.proto [hunk 23]
```diff
@@ -161,11 +161,20 @@ message Hooks {
...
+	// StartContainer is a list of hooks to be run after container is created, but before it is started.
+	repeated Hook StartContainer = 4  [(gogoproto.nullable) = false]; 
+
 	// Poststart is a list of hooks to be run after the container process is started.
-	repeated Hook Poststart = 2  [(gogoproto.nullable) = false];
+	repeated Hook Poststart = 5  [(gogoproto.nullable) = false];
 
 	// Poststop is a list of hooks to be run after the container process exits.
-	repeated Hook Poststop = 3  [(gogoproto.nullable) = false];
+	repeated Hook Poststop = 6  [(gogoproto.nullable) = false];
```
Comment (bergwolf) [inline src/libs/protocols/protos/oci.proto:177]: When editing agent protocols, please do not insert things in the middle of a struct. It will break agent protocol compatibility (which is already in a bad state.)

File: src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs [hunk 24]
```diff
@@ -57,12 +63,37 @@ impl ContainerManager for VirtContainerManager {
...
             self.agent.clone(),
             self.resource_manager.clone(),
         )
         .context("new container")?;
 
+        // CreateContainer Hooks:
+        // * should be run in vmm namespace (hook path in runtime namespace)
+        // * should be run after the vm is started, before container is created, and after CreateRuntime Hooks
+        // * spec details: https://github.com/opencontainers/runtime-spec/blob/c1662686cff159595277b79322d0272f5182941b/config.md#createcontainer-hooks
+        let vmm_master_tid = self.hypervisor.get_vmm_master_tid().await?;
```
Comment (bergwolf) [inline src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:77]: Please add a helper function to run the hook so that we don't have to maintain the same piece of code at different places.

Comment (YushuoEdge) [reply src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs:77]: I used to think the `hookstates.execute_hooks(...)` is already the helper function, since we should construct different oci::State for different kinds of hooks. I wonder, do we need to consider the hook type in the helper function? By that way, we should construct different oci::State according the hook type in that function?

File: src/libs/protocols/protos/oci.proto (hunk 23 above)
Comment (YushuoEdge) [reply src/libs/protocols/protos/oci.proto:177]: OK! I will fix it ASAP.

**Review by bergwolf — APPROVED** [review]
Thanks a lot!

File: src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs [hunk 25]
```diff
@@ -149,6 +189,16 @@ impl Sandbox for VirtSandbox {
...
 
+        // execute pre-start hook functions, including Prestart Hooks and CreateRuntime Hooks
+        let (prestart_hooks, create_runtime_hooks) = match spec.hooks.as_ref() {
+            Some(hooks) => (hooks.prestart.clone(), hooks.create_runtime.clone()),
+            None => (Vec::new(), Vec::new()),
+        };
+        self.execute_oci_hook_functions(&prestart_hooks, &create_runtime_hooks, state)
+            .await?;
+
+        // TODO: if prestart_hooks is not empty, rescan the network endpoints
```
Comment (Tim-0731-Hzt) [inline src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs:200]: Please raise an issue for this

**Review by Tim-0731-Hzt — APPROVED** [review]
LGTM

File: src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs (hunk 25 above)
Comment (YushuoEdge) [reply src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs]: Thanks! Fixed.

File: src/agent/rustjail/src/lib.rs [hunk 26]
```diff
@@ -154,12 +154,15 @@ fn hook_grpc_to_oci(h: &[grpcHook]) -> Vec<oci::Hook> {
...
     let prestart = hook_grpc_to_oci(h.Prestart.as_ref());
 
+    let create_runtime = hook_grpc_to_oci(h.CreateRuntime.as_ref());
+
     let poststart = hook_grpc_to_oci(h.Poststart.as_ref());
 
     let poststop = hook_grpc_to_oci(h.Poststop.as_ref());
 
     oci::Hooks {
         prestart,
```
Comment (c3d) [inline src/agent/rustjail/src/lib.rs:163]: I'm a bit confused. The commit message and description seem wrong: they claim that you add support for OCI hooks, but the code apparently already had `prestart`, `poststart` and `poststop`. Sho

Comment (YushuoEdge) [reply src/agent/rustjail/src/lib.rs:163]: Thanks for reviewing!

The description of this pr is actually outdated, I am sorry for that have made you confused and I have update the description.  `prestart hooks`, `poststart hooks`, `poststop hooks` are those hooks that have been supported in runtime-go, so I guess this is the reason why the code already have `prestart`, `poststart`, and `poststop`. And `create_runtime hooks` will not be executed in the guest, so I guess this is the reason why it does not appear in the original code.

But actually, `prestart`, `poststart` and `poststop` *OCI hooks* will not execute in the guest either, the reason I tend to keep them here is for better log output and for future changes. WDYT?

I hope my explanation can make you clean😆

Comment (c3d) [reply src/agent/rustjail/src/lib.rs:163]: I think that the new wording is much better, thank you @YushuoEdge

File: docs/design/hooks-handling.md [hunk 27]
```diff
@@ -0,0 +1,62 @@
+# Kata Containers support for `Hooks`
```
Comment (c3d) [inline docs/design/hooks-handling.md:1]: If you do a new commit for any other reason, I would suggest reordering your commits to put this one first.

File: src/agent/rustjail/src/lib.rs [hunk 28]
```diff
@@ -860,6 +864,7 @@ mod tests {
                         env: Vec::from([String::from("env1"), String::from("env2")]),
                         timeout: Some(10),
                     }]),
+                    ..Default::default()
```
Comment (c3d) [inline src/agent/rustjail/src/lib.rs:867]: I find the use of `default` here inconsistent with how it's done for other hooks. Could you either:

1. Expand the initialization like for the other hooks, or
2. Explain in a comment why defaults are OK for the newly created hooks.

File: src/libs/protocols/src/trans.rs [hunk 29]
```diff
@@ -974,6 +977,18 @@ impl From<crate::oci::Hooks> for oci::Hooks {
         for hook in from.take_Prestart().to_vec() {
             prestart.push(hook.into())
         }
+        let mut create_runtime = Vec::new();
```
Comment (c3d) [inline src/libs/protocols/src/trans.rs:980]: While this is similar to the code that existed before, this is not super idiomatic. You could use something like:

```
        let create_container = from
            .take_CreateContainer()
            .into_iter()
            .map(|i| i.into())
            .collect();
```

This dispense with the need to make things `mut`.

File: docs/design/hooks-handling.md (hunk 27 above)
Comment (YushuoEdge) [reply docs/design/hooks-handling.md:1]: Update!

File: src/agent/rustjail/src/lib.rs (hunk 28 above)
Comment (YushuoEdge) [reply src/agent/rustjail/src/lib.rs]: I have expanded the initialization!

File: src/libs/protocols/src/trans.rs (hunk 29 above)
Comment (YushuoEdge) [reply src/libs/protocols/src/trans.rs]: Great suggestion! Thanks!

**Review by studychao — APPROVED** [review]
Really nice work. Thanks @YushuoEdge
